components/dashboards/org.wso2.carbon.dashboards.api/src/main/java/org/wso2/carbon/dashboards/api/internal/DashboardRestApi.java

```diff
@@ -182,4 +186,66 @@ public Response delete(@PathParam("id") String id) {
             return Response.serverError().entity("Cannot delete dashboard '" + id + "'.").build();
         }
     }
+
+    @GET
+    @Path("/roles")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getRoles() {
+        try {
+            List<Role> allRoles = dashboardDataProvider.getAllRoles();
+            return Response.ok()
+                    .entity(allRoles)
+                    .type(MediaType.APPLICATION_JSON)
+                    .build();
+        } catch (DashboardException e) {
+            LOGGER.error("An error occurred while retrieving roles.", e);
+            return Response.serverError()
+                    .entity("An error occurred while retrieving roles.")
+                    .build();
+        }
+    }
+
+    @GET
+    @Path("/roles/{username}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getRolesByUsername(@PathParam("username") String username) {
+        try {
+            List<Role> roles = dashboardDataProvider.getRolesByUsername(username);
+            return Response.ok()
+                    .entity(roles)
+                    .type(MediaType.APPLICATION_JSON)
+                    .build();
+        } catch (DashboardException e) {
+            LOGGER.error("An error occurred while retrieving user roles.", e);
+            return Response.serverError()
+                    .entity("An error occurred while retrieving user roles.")
+                    .build();
+        }
+    }
+
+    @GET
+    @Path("/{url}/roles")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getDashboardRoles(@PathParam("url") String url) {
+        return Response.ok()
+                .entity(dashboardDataProvider.getDashboardRoles(url))
+                .type(MediaType.APPLICATION_JSON)
+                .build();
+    }
+
+    @POST
+    @Path("/{url}/roles")
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response updateDashboardRoles(@PathParam("url") String url, Map<String, List<String>> roles) {
+        try {
+            dashboardDataProvider.updateDashboardRoles(url, roles);
+            return Response.ok().build();
+        } catch (DashboardException e) {
+            LOGGER.error("An error occurred while updating dashboard roles.", e);
+            return Response.serverError()
+                    .entity("An error occurred while updating dashboard roles.")
```

Comment:
error message: `Cannot update user roles of dashboard ' + url + '.`

Reply:
Ack